src/workerd/api/tracing-module.c++

```diff
@@ -0,0 +1,225 @@
+// Copyright (c) 2017-2025 Cloudflare, Inc.
+// Licensed under the Apache 2.0 license found in the LICENSE file or at:
+//     https://opensource.org/licenses/Apache-2.0
+
+#include "tracing-module.h"
+
+#include "src/workerd/jsg/inspector.h"
+
+#include <workerd/api/actor-state.h>
+#include <workerd/api/global-scope.h>
+
+namespace workerd::api {
+
+// InternalSpan implementation
+InternalSpan::InternalSpan(kj::Maybe<IoOwn<InternalSpanImpl>> impl): impl(kj::mv(impl)) {}
+
+void InternalSpan::end() {
+  KJ_IF_SOME(s, impl) {
+    s->end();
+  }
+}
+
+void InternalSpan::setTag(
+    jsg::Lock& js, kj::String key, jsg::Optional<kj::OneOf<bool, double, kj::String>> maybeValue) {
+  KJ_IF_SOME(value, maybeValue) {
+    kj::OneOf<bool, double, kj::String> tagValue;
+
+    KJ_SWITCH_ONEOF(value) {
+      KJ_CASE_ONEOF(s, kj::String) {
+        tagValue = kj::mv(s);
+      }
+      KJ_CASE_ONEOF(b, bool) {
+        tagValue = b;
+      }
+      KJ_CASE_ONEOF(d, double) {
+        tagValue = d;
+      }
+    }
+
+    KJ_IF_SOME(s, impl) {
+      s->setTag(kj::ConstString(kj::mv(key)), kj::mv(tagValue));
+    }
+  }
+}
+
+bool InternalSpan::getIsRecording() {
+  KJ_IF_SOME(s, impl) {
+    return s->getIsRecording();
+  }
+  return false;
+}
+
+// InternalSpanImpl implementation
+InternalSpanImpl::InternalSpanImpl(SpanBuilder span): span(kj::mv(span)) {}
+
+InternalSpanImpl::~InternalSpanImpl() noexcept(false) {
+  end();
+}
+
+void InternalSpanImpl::end() {
+  KJ_IF_SOME(s, span) {
+    s.end();
+    span = kj::none;
```

Comment:
Note: Deallocating the span doesn't really do anything if we have already called end. Based on that, we could simplify InternalSpanImpl by removing the kj::Maybe from span (assuming that doesn't end up interfering with getIsRecording(), but I think it's possible if we make InternalSpan::end() deallocate its InternalSpanImpl).